src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,389 @@
+			menu.scrollTop = element.offsetTop;
+		}
+		// Scroll menu down so bottom of highlighted result aligns with bottom of menu container
+		else if ((element.offsetTop - menu.scrollTop + element.offsetHeight) > menu.clientHeight) {
+			menu.scrollTop = element.offsetTop - menu.clientHeight + element.offsetHeight;
+		}
+	}
+
+	onInput(event: Event) {
+		const {
+			onRequestResults,
+			onChange
+		} = this.properties;
+
+		this.updateSelectedIndex(Operation.reset);
+		this._open = true;
+		onChange && onChange((<HTMLInputElement> event.target).value);
+		onRequestResults && onRequestResults((<HTMLInputElement> event.target).value);
+	}
+
+	onInputFocus(event: FocusEvent) {
+		const {
+			openOnFocus,
+			onFocus,
+			onRequestResults
+		} = this.properties;
+
+		const value = (<HTMLInputElement> event.target).value;
+
+		this._focused = true;
+		this._ignoreBlur = false;
+		onFocus && onFocus(value);
+
+		if (openOnFocus && !this._ignoreFocus) {
+			this._open = true;
+			onRequestResults && onRequestResults(value);
+		}
+
+		this._ignoreFocus = false;
+	}
+
+	onInputBlur(event: FocusEvent) {
+		const { onBlur } = this.properties;
+
+		if (this._ignoreBlur) {
+			return;
+		}
+
+		this.updateSelectedIndex(Operation.reset);
+		this._open = false;
+		this._focused = false;
+		onBlur && onBlur((<HTMLInputElement> event.target).value);
+		this.invalidate();
+	}
+
+	onInputKeyDown(event: KeyboardEvent) {
+		this._handlers[event.keyCode] && this._handlers[event.keyCode].call(this, event);
+	}
+
+	onArrowClick() {
+		const {
+			onRequestResults,
+			disabled,
+			readOnly
+		} = this.properties;
+
+		if (disabled || readOnly) {
+			return;
+		}
+
+		this._open = true;
+		this._focused = true;
+		this._inputElement && onRequestResults && onRequestResults(this._inputElement.value);
+	}
+
+	onClearClick() {
+		const { onChange } = this.properties;
+		this._focused = true;
+		onChange && onChange('');
+	}
+
+	onResultMouseEnter(index: number) {
+		this.updateSelectedIndex(undefined, index);
+		this.invalidate();
+	}
+
+	onResultMouseDown() {
+		this._ignoreBlur = true;
+	}
+
+	onResultMouseUp() {
+		const { results } = this.properties;
+
+		if (!results || this._selectedIndex === undefined) {
+			return;
+		}
+
+		this.selectResult(results[this._selectedIndex]);
+	}
+
+	selectResult(result: any) {
+		const {
+			autoBlur,
+			onChange
+		} = this.properties;
+
+		this.updateSelectedIndex(Operation.reset);
+		this._open = false;
+		this._focused = !autoBlur;
+		this._ignoreFocus = true;
+		onChange && onChange(this.getResultLabel(result));
+	}
+
+	getResultLabel(result: any): string {
+		const { getResultLabel } = this.properties;
+		return getResultLabel ? getResultLabel(result) : result;
+	}
+
+	renderMenu(results: any[]) {
+		const menu = w('result-menu', <ResultMenuProperties> {
+			registry: this.registry,
+			results: results,
+			selectedIndex: this._selectedIndex,
+			getResultLabel: this.getResultLabel,
+			skipResult: () => {
```

Comment:
This will cause a property change to be detected every render.

could you add a method to the class?